File: design_patterns/01_Singleton/Logger.java
import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;

public class Logger {
    private static Logger instance;
    private static final Object lock = new Object();
    private final String logFile = "app.log";

    private Logger() {
        // private constructor to prevent instantiation
    }

    public static Logger getInstance() {
        if (instance == null) {
            synchronized (lock) {
                if (instance == null) {
                    instance = new Logger();
                }
            }
        }
        return instance;
    }

    public void log(String message) {
        try (FileWriter fw = new FileWriter(logFile, true);
             PrintWriter pw = new PrintWriter(fw)) {
            pw.println(message);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    // Example usage
    public static void main(String[] args) {
        Logger logger1 = Logger.getInstance();
        Logger logger2 = Logger.getInstance();

        logger1.log("Logging from logger1");
        logger2.log("Logging from logger2");

        System.out.println("Logger1 and Logger2 are the same instance: " + (logger1 == logger2));
    }
}

File: design_patterns/01_Singleton/logger.js
const fs = require('fs');
const path = require('path');

class Logger {
    constructor() {
        if (Logger.instance) {
            return Logger.instance;
        }
        
        this._logFile = path.join(__dirname, 'app.log');
        Logger.instance = this;
    }

    log(message) {
        fs.appendFileSync(this._logFile, `${message}\n`);
    }
}

// Example usage
const logger1 = new Logger();
const logger2 = new Logger();

logger1.log("Logging from logger1");
logger2.log("Logging from logger2");

console.log(`Logger1 and Logger2 are the same instance: ${logger1 === logger2}`);
